migration/src/main/scala/akka/persistence/r2dbc/migration/DurableStateMigrationToolDao.scala

```diff
@@ -0,0 +1,66 @@
+/*
+ * Copyright (C) 2022 - 2023 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package akka.persistence.r2dbc.migration
+
+import java.time.Instant
+
+import scala.concurrent.ExecutionContext
+import scala.concurrent.Future
+
+import akka.Done
+import akka.annotation.InternalApi
+import akka.dispatch.ExecutionContexts
+import akka.persistence.r2dbc.internal.Dialect
+import akka.persistence.r2dbc.internal.DurableStateDao.SerializedStateRow
+import akka.persistence.r2dbc.internal.R2dbcExecutor
+import akka.persistence.r2dbc.internal.R2dbcExecutorProvider
+import akka.persistence.r2dbc.internal.postgres.PostgresDurableStateDao
+import akka.persistence.typed.PersistenceId
+
+/**
+ * INTERNAL API
+ */
+@InternalApi private[r2dbc] class DurableStateMigrationToolDao(
+    executorProvider: R2dbcExecutorProvider,
+    dialect: Dialect)(implicit ec: ExecutionContext)
+    extends PostgresDurableStateDao(executorProvider, dialect) {
+
+  def upsertDurableState(state: SerializedStateRow, value: Any): Future[Done] = {
+    val slice = persistenceExt.sliceForPersistenceId(state.persistenceId)
+    val r2dbcExecutor = executorProvider.executorFor(slice)
+    val entityType = PersistenceId.extractEntityType(state.persistenceId)
+    readState(state.persistenceId).flatMap {
+      case Some(existingState) => update(r2dbcExecutor, existingState, state, value, slice, entityType)
+      case None                => insert(r2dbcExecutor, state, value, slice, entityType)
+    }
```

Comment:
Wouldn't it be better to leverage the upsert support in Postgres, instead of not having to do a read roundtrip for every migrated state write?

Reply:
Yes, I had the same idea and Patrik [suggested](https://github.com/akka/akka-persistence-r2dbc/pull/553) the same. But when I looked at it, it seemed that an upsert can only decide on the exact primary key match, which is not what we want here I think.

Reply:
I see, we have both pid and revision as part of primary key, I didn't remember that. Probably no way around it then. 👍

Reply:
wow, `PRIMARY KEY(persistence_id, revision)` must be an oversight. We don't ever want to store more than one row for a persistence_id. The application logic takes care of it, but feels wrong. Created ticket https://github.com/akka/akka-persistence-r2dbc/issues/558 for separate follow up.